File: merged.go
// Copyright 2019 Google Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package reftable

import (
	"bytes"
	"fmt"
	"log"
	"sort"
)

// pqEntry is an entry of the priority queue
type pqEntry struct {
	rec   Record
	index int
}

func (e pqEntry) String() string {
	return fmt.Sprintf("%s,%d", e.rec.Key(), e.index)
}

func pqLess(a, b pqEntry) bool {
	if a.rec.Key() == b.rec.Key() {
		return a.index > b.index
	}

	return a.rec.Key() < b.rec.Key()
}

// a min heap
type mergedIterPQueue struct {
	heap []pqEntry
}

func (pq *mergedIterPQueue) isEmpty() bool {
	return len(pq.heap) == 0
}

func (pq *mergedIterPQueue) top() pqEntry {
	return pq.heap[0]
}

func (pq *mergedIterPQueue) check() {
	for i := 1; i < len(pq.heap); i++ {
		par := (i - 1) / 2

		if !pqLess(pq.heap[par], pq.heap[i]) {
			log.Panicf("%v %v %d", pq.heap[par], pq.heap[i], i)
		}
	}
}

func (pq *mergedIterPQueue) remove() pqEntry {
	e := pq.heap[0]

	pq.heap[0] = pq.heap[len(pq.heap)-1]
	pq.heap = pq.heap[:len(pq.heap)-1]

	i := 0
	for i < len(pq.heap) {
		min := i
		j, k := 2*i+1, 2*i+2

		if j < len(pq.heap) && pqLess(pq.heap[j], pq.heap[i]) {
			min = j
		}
		if k < len(pq.heap) && pqLess(pq.heap[k], pq.heap[min]) {
			min = k
		}

		if min == i {
			break
		}

		pq.heap[i], pq.heap[min] = pq.heap[min], pq.heap[i]
		i = min
	}

	return e
}

func (pq *mergedIterPQueue) add(e pqEntry) {
	pq.heap = append(pq.heap, e)
	i := len(pq.heap) - 1
	for i > 0 {
		j := (i - 1) / 2

		if pqLess(pq.heap[j], pq.heap[i]) {

			break
		}

		pq.heap[j], pq.heap[i] = pq.heap[i], pq.heap[j]
		i = j
	}
}

// Merged is a stack of reftables.
type Merged struct {
	stack []*Reader
}

// NewMerged creates a reader for a merged reftable.
func NewMerged(tabs []*Reader) (*Merged, error) {
	var last Table
	for i, t := range tabs {
		if last != nil && last.MaxUpdateIndex() >= t.MinUpdateIndex() {
			return nil, fmt.Errorf("reftable: table %d has min %d, table %d has max %d; indices must be increasing.", i, t.MinUpdateIndex(), i-1, last.MaxUpdateIndex())
		}
		last = t
	}

	return &Merged{tabs}, nil
}

// MaxUpdateIndex implements the Table interface.
func (m *Merged) MaxUpdateIndex() uint64 {
	return m.stack[len(m.stack)].MaxUpdateIndex()
}

// MinUpdateIndex implements the Table interface.
func (m *Merged) MinUpdateIndex() uint64 {
	return m.stack[0].MinUpdateIndex()
}

// filteringRefIterator applies filtering for `oid` to the block
type filteringRefIterator struct {
	// The object ID to filter for
	oid []byte

	// doubleCheck if set, will cause the refs to be checked
	// against the table in `tab`.
	doubleCheck bool
	tab         Table

	// mutable
	it Iterator
}

// Next implements the Iterator interface.
func (it *filteringRefIterator) Next(rec Record) (bool, error) {
	ref := rec.(*RefRecord)
	for {
		ok, err := it.it.Next(ref)
		if !ok || err != nil {
			return false, err
		}

		if it.doubleCheck {
			it, err := it.tab.Seek(ref)
			if err != nil {
				return false, err
			}

			ok, err := it.Next(ref)
			if !ok || err != nil {
				return false, err
			}
		}

		if bytes.Compare(ref.Value, it.oid) == 0 || bytes.Compare(ref.TargetValue, it.oid) == 0 {
			return true, err
		}
	}
}

// RefsFor returns the refs that point to the given oid
func (m *Merged) RefsFor(oid []byte) (Iterator, error) {
	mit := &mergedIter{
		typ: BlockTypeRef,
	}
	for _, t := range m.stack {
		it, err := t.RefsFor(oid)
		if err != nil {
			return nil, err
		}
		mit.stack = append(mit.stack, it)
	}

	if err := mit.init(); err != nil {
		return nil, err
	}
	return &filteringRefIterator{
		tab:         m,
		oid:         oid,
		it:          mit,
		doubleCheck: true,
	}, nil
}

func uniq(ss []string) []string {
	sort.Strings(ss)
	u := ss[:0]
	last := ""
	for i, s := range ss {
		if i == 0 || last != s {
			u = append(u, s)
		}

		last = s
	}
	return u
}

// Seek returns an iterator positioned before the wanted record.
func (m *Merged) Seek(rec Record) (Iterator, error) {
	var its []Iterator
	for _, t := range m.stack {
		iter, err := t.Seek(rec)
		if err != nil {
			return nil, err
		}
		its = append(its, iter)
	}

	merged := &mergedIter{
		typ:   rec.Type(),
		stack: its,
	}

	if err := merged.init(); err != nil {
		return nil, err
	}

	return merged, nil
}

// mergedIter iterates over a stack of reftables. Entries from higher
// in the stack obscure lower entries.
type mergedIter struct {
	// XXX - could short-circuit the entries coming from the base stack
	typ   byte
	pq    mergedIterPQueue
	stack []Iterator
}

func (it *mergedIter) init() error {
	for i, sub := range it.stack {
		rec := newRecord(it.typ, "")
		ok, err := sub.Next(rec)
		if err != nil {
			return err
		}
		if ok {
			it.pq.add(pqEntry{
				rec:   rec,
				index: i,
			})
		} else {
			it.stack[i] = nil
		}
	}

	return nil
}

// advance one iterator, putting its result into the priority queue.
func (m *mergedIter) advanceSubIter(index int) error {
	if m.stack[index] == nil {
		return nil
	}

	r := newRecord(m.typ, "")
	ok, err := m.stack[index].Next(r)
	if err != nil {
		return err
	}

	if !ok {
		m.stack[index] = nil
		return nil
	}

	m.pq.add(pqEntry{r, index})
	return nil
}

func (m *mergedIter) Next(rec Record) (bool, error) {
	for {
		ok, err := m.next(rec)
		if !ok || err != nil {
			return ok, err
		}
		return true, nil
	}
}

func (m *mergedIter) next(rec Record) (bool, error) {
	if m.pq.isEmpty() {
		return false, nil
	}

	entry := m.pq.remove()
	m.advanceSubIter(entry.index)

	for !m.pq.isEmpty() {
		top := m.pq.top()
		if top.rec.Key() > entry.rec.Key() {
			break
		}

		m.pq.remove()
		m.advanceSubIter(top.index)
	}

	rec.CopyFrom(entry.rec)
	return true, nil
}

File: c/merged.c
// Copyright 2019 Google Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <stdlib.h>

#include "iter.h"
#include "pq.h"
#include "reader.h"

struct merged_table {
  struct reader **stack;
  int stack_len;

  uint64_t min;
  uint64_t max;
};

struct merged_iter {
  struct iterator *stack;
  int stack_len;
  byte typ;
  struct merged_iter_pqueue pq;
} merged_iter;

int merged_iter_init(struct merged_iter *mi) {
  for (int i = 0; i < mi->stack_len; i++) {
    struct record rec = new_record(mi->typ);
    int err = iterator_next(mi->stack[i], rec);
    if (err < 0) {
      return err;
    }

    if (err > 0) {
      iterator_destroy(&mi->stack[i]);
    } else {
      struct pq_entry e = {
          .rec = rec,
          .index = i,
      };
      merged_iter_pqueue_add(&mi->pq, e);
    }
  }

  return 0;
}

void merged_iter_close(void *p) {
  struct merged_iter *mi = (struct merged_iter *)p;

  merged_iter_pqueue_clear(&mi->pq);
  for (int i = 0; i < mi->stack_len; i++) {
    iterator_destroy(&mi->stack[i]);
  }
}

int merged_iter_advance_subiter(struct merged_iter *mi, int idx) {
  if (iterator_is_null(mi->stack[idx])) {
    return 0;
  }

  struct record rec = new_record(mi->typ);
  int err = iterator_next(mi->stack[idx], rec);
  if (err < 0) {
    return err;
  }

  if (err > 0) {
    iterator_destroy(&mi->stack[idx]);
    return 0;
  }

  struct pq_entry e = {
      .rec = rec,
      .index = idx,
  };
  merged_iter_pqueue_add(&mi->pq, e);
  return 0;
}

int merged_iter_next(void *p, struct record rec) {
  struct merged_iter *mi = (struct merged_iter *)p;
  if (merged_iter_pqueue_is_empty(mi->pq)) {
    return 1;
  }

  struct pq_entry entry = merged_iter_pqueue_remove(&mi->pq);
  int err = merged_iter_advance_subiter(mi, entry.index);
  if (err < 0) {
    return err;
  }

  struct slice entry_key = {};
  record_key(entry.rec, &entry_key);
  while (!merged_iter_pqueue_is_empty(mi->pq)) {
    struct pq_entry top = merged_iter_pqueue_top(mi->pq);

    struct slice k = {};
    record_key(top.rec, &k);

    int cmp = slice_compare(k, entry_key);
    free(slice_yield(&k));

    if (cmp > 0) {
      break;
    }

    merged_iter_pqueue_remove(&mi->pq);
    int err = merged_iter_advance_subiter(mi, top.index);
    if (err < 0) {
      return err;
    }
    record_clear(top.rec);
    free(record_yield(&top.rec));
  }

  record_copy_from(rec, entry.rec);
  record_clear(entry.rec);
  free(record_yield(&entry.rec));
  free(slice_yield(&entry_key));
  return 0;
}

struct iterator_ops merged_iter_ops = {
    .next = &merged_iter_next,
    .close = &merged_iter_close,
};

void iterator_from_merged_iter(struct iterator *it, struct merged_iter *mi) {
  it->iter_arg = mi;
  it->ops = &merged_iter_ops;
}

/* new_merged_table creates a new merged table. It takes ownership of the stack
 * array. */
int new_merged_table(struct merged_table **dest, struct reader **stack, int n) {
  uint64_t last_max = 0;
  uint64_t first_min = 0;
  for (int i = 0; i < n; i++) {
    struct reader *r = stack[i];
    if (i > 0 && last_max >= reader_min_update_index(r)) {
      return FORMAT_ERROR;
    }
    if (i == 0) {
      first_min = reader_min_update_index(r);
    }

    last_max = reader_max_update_index(r);
  }

  struct merged_table m = {
      .stack = stack,
      .stack_len = n,
      .min = first_min,
      .max = last_max,
  };

  *dest = calloc(sizeof(struct merged_table), 1);
  **dest = m;
  return 0;
}

uint64_t merged_max_update_index(struct merged_table *mt) { return mt->max; }

uint64_t merged_min_update_index(struct merged_table *mt) { return mt->min; }

int merged_table_seek_record(struct merged_table *mt, struct iterator *it,
                             struct record rec) {
  struct iterator *iters = calloc(sizeof(struct iterator), mt->stack_len);
  for (int i = 0; i < mt->stack_len; i++) {
    int err = reader_seek(mt->stack[i], &iters[i], rec);
    if (err < 0) {
      // XXX leak.
      return err;
    }
  }

  struct merged_iter merged = {
      .stack = iters,
      .stack_len = mt->stack_len,
      .typ = record_type(rec),
  };

  int err = merged_iter_init(&merged);
  if (err < 0) {
    return err;
  }

  struct merged_iter *p = malloc(sizeof(struct merged_iter));
  *p = merged;
  iterator_from_merged_iter(it, p);
  return 0;
}

int merged_table_seek_ref(struct merged_table *mt, struct iterator *it,
                          struct ref_record *ref) {
  struct record rec = {};
  record_from_ref(&rec, ref);
  return merged_table_seek_record(mt, it, rec);
}